test/backend.cpp

```diff
@@ -8,22 +8,51 @@
  ********************************************************/
 
 #include <arrayfire.h>
-#include <gtest/gtest.h>
-#include <testHelpers.hpp>
+#include <af/backend.h>
 #include <af/data.h>
+#include <af/device.h>
 #include <af/dim4.hpp>
 #include <af/traits.hpp>
+
+#include <gtest/gtest.h>
+#include <testHelpers.hpp>
+
 #include <string>
 #include <vector>
 
-#include <af/device.h>
-
+using af::array;
+using af::Backend;
 using af::dtype_traits;
+using af::exception;
 using af::getAvailableBackends;
+using af::getBackendCount;
+using af::randu;
 using af::setBackend;
+using af::transpose;
 using std::string;
 using std::vector;
 
+// These paths are based on where the CI configuration puts the built libraries
+const string build_dir_str = BUILD_DIR;
+#if defined(_WIN32)
+const string library_suffix = ".dll";
```

Comment:
I would suggest passing these as definitions via CMake, library suffix values are already available as variables in CMake.

Reply:
You mean if the location of the binaries change from e.g. `build/src/backend/cpu`, correct? The CMake route makes sense, I'll try that.

Reply:
Yes, instead of something like `<path/libafcpu.so` we just give `<path>` that may contain one or more binaries of our backends. Any and all available backends (afcpu, afcuda, afopencl) will be loaded.

Having said that, now I realize this would change the logic completely. Just throwing out an idea. In this approach, the user code will reduce from 3 lines to 1 line to add all three(cpu,cuda,opencl) new backends. If we add more backends, just saying that loading all in one line is better having a line for each backend we want to load.

Of course, in this new approach, we would have to provide a new function that prints/logs all available backends - much similar to how we list our devices in each backend - so that later the user can call `af_set_backend_library` on required index. We can have deterministic order in which backend handles are loaded so that cuda -> opencl -> cpu in a given path.

Reply:
So I see two concerns here, the first being the original concern that Umar raised that I think we should solve first.

First, these kinds of variables are fragile:
```cpp
#elif defined(__linux__)
const string library_suffix = ".so";
const string library_prefix_cpu = "/src/backend/cpu/libafcpu";
const string library_prefix_cuda = "/src/backend/cuda/libafcuda";
const string library_prefix_opencl = "/src/backend/opencl/libafopencl";
```
Because these locations might change in the future - I agree. I doubt that the library names would change though? But it's true that it would be better if even the library names are gotten from CMake. I would say the same for the file extensions.

The second concern is possibly providing the directory path that contains the libraries to be added. I agree that it will change the logic for `af_set_backend_library` indeed (about how to determine the index that should be used to activate the desired library), and I agree with @9prady9's thoughts in how that might possibly play out. I think though that this concern can be considered instead in the future, and we should just consider this basic functionality for now.

Reply:
Just a thought - the probability of library name change is about the same as those build directory change. File extensions on the other hand, I doubt `dll` or `so` or `dylib` are going to change in our lifetime :100:

Reply:
Thanks, I agree. I think it's still better if these strings/paths were formed by CMake as you guys have suggested though. The way I handle this now is by checking the value of `CMAKE_RUNTIME_OUTPUT_DIRECTORY` first. If it's non-empty (as is the case in Windows builds), then I use that path for finding the libraries. Else (if empty), then that means it was not set in any CMake configuration file, and thus the [output directory will default to the corresponding source directory hierarchy in the build tree](https://cmake.org/cmake/help/latest/command/add_library.html) (which in our case is `src/backend/cpu` and the like). I made those paths into variables and use them in the tests' CMakeLists. Here's the proposed change:

```cmake
# <arrayfire_root>/CMakeLists.txt
set(CPU_BACKEND_SOURCE_DIR     src/backend/cpu)
set(CUDA_BACKEND_SOURCE_DIR    src/backend/cuda)
set(OPENCL_BACKEND_SOURCE_DIR  src/backend/opencl)
set(UNIFIED_BACKEND_SOURCE_DIR src/api/unified)

conditional_directory(AF_BUILD_CPU     ${CPU_BACKEND_SOURCE_DIR})
conditional_directory(AF_BUILD_CUDA    ${CUDA_BACKEND_SOURCE_DIR})
conditional_directory(AF_BUILD_OPENCL  ${OPENCL_BACKEND_SOURCE_DIR})
conditional_directory(AF_BUILD_UNIFIED ${UNIFIED_BACKEND_SOURCE_DIR})

# test/CMakeLists.txt
set(CPU_LIB_NAME "${CMAKE_SHARED_LIBRARY_PREFIX}afcpu${CMAKE_SHARED_LIBRARY_SUFFIX}")
set(CUDA_LIB_NAME "${CMAKE_SHARED_LIBRARY_PREFIX}afcuda${CMAKE_SHARED_LIBRARY_SUFFIX}")
set(OPENCL_LIB_NAME "${CMAKE_SHARED_LIBRARY_PREFIX}afopencl${CMAKE_SHARED_LIBRARY_SUFFIX}")

if(CMAKE_RUNTIME_OUTPUT_DIRECTORY)
  set(CPU_LIB_PATH "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${CPU_LIB_NAME}")
  set(CUDA_LIB_PATH "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${CUDA_LIB_NAME}")
  set(OPENCL_LIB_PATH "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${OPENCL_LIB_NAME}")
else()
  set(CPU_LIB_PATH "${ArrayFire_BINARY_DIR}/${CPU_BACKEND_SOURCE_DIR}/${CPU_LIB_NAME}")
  set(CUDA_LIB_PATH "${ArrayFire_BINARY_DIR}/${CUDA_BACKEND_SOURCE_DIR}/${CUDA_LIB_NAME}")
  set(OPENCL_LIB_PATH "${ArrayFire_BINARY_DIR}/${OPENCL_BACKEND_SOURCE_DIR}/${OPENCL_LIB_NAME}")
endif()

add_definitions("-DCPU_LIB_PATH=\"${CPU_LIB_PATH}\"")
add_definitions("-DCUDA_LIB_PATH=\"${CUDA_LIB_PATH}\"")
add_definitions("-DOPENCL_LIB_PATH=\"${OPENCL_LIB_PATH}\"")
```